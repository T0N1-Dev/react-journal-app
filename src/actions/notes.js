import { addDoc, collection, doc, updateDoc, deleteDoc } from "firebase/firestore"
import { db } from "../firebase/firebaseConfig"
import { types } from "../types/types"
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = () => {
    return async ( dispatch, getState ) => {

        const uid = getState().auth.uid;
        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime(),
        }

        Swal.fire({
            title: 'Adding...',
            text: 'Please wait...', 
            allowOutsideClick: false,
            didOpen: () => {
                Swal.showLoading();
            }
        });
        
        try {
            const docRef = await addDoc(collection(db, `${uid}/journal/notes`), newNote);
            const noteWithId = { id: docRef.id, ...newNote };
            dispatch( activeNote( docRef.id, newNote ));
            dispatch(addNoteToState(noteWithId));
        } catch (error) {
            Swal.fire('Error al añadir la nueva nota:', error, 'error');
        } finally {
            Swal.close();
        }
    }
};

export const activeNote = ( id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const addNoteToState = (note) => ({
    type: types.noteAdded,
    payload: note
});

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes( uid );
        dispatch(setNotes(notes));
    }
}

export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes
});

export const startSaveNote = ( note ) => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        const noteToFirestore = { ...note };
        delete noteToFirestore.id;

        try {
            const notRef = doc(db, `${uid}/journal/notes/${note.id}`);
            await updateDoc(notRef, noteToFirestore);
            dispatch( refreshNote(note.id, noteToFirestore ));
            dispatch( activeNote(note.id, noteToFirestore));
            Swal.fire('Nota actualizada correctamente', note.title, 'success');
        } catch (error) {
            Swal.fire('Error al actualizar la nota:', error, 'error');
        }
    }
}

export const refreshNote = ( id, note ) => ({
    type: types.notesUpdated,
    payload: {
        id, 
        note: {
            id,
            ...note
        }
    }
});

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active: activeNote } = getState().notes;

        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...', 
            allowOutsideClick: false,
            didOpen: () => {
                Swal.showLoading();
            }
            
        });

        try {
            const fileUrl = await fileUpload( file );

            const updatedNote = {
                ...activeNote,
                url: fileUrl
            };
            dispatch( startSaveNote( updatedNote ));
        } catch (error) {
            Swal.fire('Error', 'Failed to upload the file', 'error');
        } finally {
            Swal.close();
        }
    }
}

export const startDeleting = ( id ) => {
    return (dispatch, getState) => {
         const { uid } = getState().auth;
         const notRef = doc(db, `${uid}/journal/notes/${id}`);
         
         try {
            Swal.fire({
                title: "Are you sure?",
                text: "You won't be able to revert this!",
                icon: "warning",
                showCancelButton: true,
                confirmButtonColor: "#3085d6",
                cancelButtonColor: "#d33",
                confirmButtonText: "Yes, delete it!"
            }).then(async (result) => {
                if (result.isConfirmed) {
                    Swal.fire({
                        timer: 1000,
                        timerProgressBar: true,
                        allowOutsideClick: false,
                        didOpen: () => {
                            Swal.showLoading();
                        }
                    });
                    await deleteDoc(notRef);
                    dispatch( deleteNote(id) );
                    Swal.fire({
                        title: "Deleted!",
                        text: "Your note has been deleted.",
                        icon: "success"
                    });
                }
              });
        } catch (e) {
            Swal.fire('Ha ocurrido un error durante la eliminación de la nota', e, 'error');
        }
         
    };
};

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
});

export const noteLogout = () => ({
    type: types.notesLogoutCleaning
})
