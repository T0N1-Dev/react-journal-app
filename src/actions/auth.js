import { types } from "../types/types";
import { createUserWithEmailAndPassword, signInWithEmailAndPassword, signInWithPopup, updateProfile } from "firebase/auth";
import { auth, googleAuthProvider } from "../firebase/firebaseConfig";
import { finishLoading, startLoading } from "./ui";
import Swal from 'sweetalert2';
import { noteLogout } from "./notes";

export const startLoginEmailPassword = (email, password) => {

    return async (dispatch) => {
        try {
            dispatch(startLoading());
            const {user} = await signInWithEmailAndPassword(auth, email, password);

            dispatch( login(user.uid, user.displayName));
            dispatch(finishLoading());
        }
        catch (err) {
            dispatch(finishLoading());
            Swal.fire('Error', err.message, 'error');
        } 
    }
}

export const startRegisterWithEmailPassswordName = (email, password, name) => {

    return async (dispatch) => {
        try {
            dispatch(startLoading());
            const { user } = await createUserWithEmailAndPassword(auth, email, password);
            await updateProfile(user, { displayName: name });
            
            dispatch(
                login(user.uid, user.displayName)
            );
            dispatch(finishLoading());
        }
        catch (err) {
            Swal.fire('Error', err.message, 'error');
            dispatch(finishLoading())
        } 
    }
}

export const startGoogleLogin = () => {
    return async (dispatch) => {
        try {
          const { user } = await signInWithPopup(auth, googleAuthProvider);
          dispatch(
            login( user.uid, user.displayName )
          );
        } catch (error) {
          console.error("Error al iniciar sesiÃ³n con Google:", error);
        }
    };
}

export const login = (uid, displayName) => ({
        type: types.login,
        payload: {
            uid,
            displayName
        }
    })

export const startLogout = () => {
    return async (dispatch) => {
        try {
            await auth.signOut();
            dispatch( logout() );
            dispatch( noteLogout() );
        } catch (err) {
            console.error("Error al hacer Logout:", err);
        }
    }
}

export const logout = () => ({
    type: types.logout
})