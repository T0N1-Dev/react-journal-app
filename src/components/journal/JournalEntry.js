import moment from 'moment'
import React from 'react'
import { activeNote } from '../../actions/notes';
import { useDispatch } from 'react-redux';

export const JournalEntry = ({id, date, title, body, url}) => {

    const noteDate = moment(date);
    const dispatch = useDispatch();

    const handleEntryClick = () => {
        dispatch(activeNote(id, {title, body, date, url}));
    }

  return (
    <div 
        className='journal__entry pointer animate__animated animate__backInLeft animate__faster'
        onClick={handleEntryClick}
    >

        {
            url &&
            <div 
                className='journal_entry-picture'
                style={{
                    backgroundSize: 'cover',
                    backgroundImage: `url(${url})`
                }}
            ></div>
        }

        <div className='journal_entry-body'>
            <p className='journal__entry-title'>
                {title}
            </p>
            <p className='journal__entry-content'>
                {body}
            </p>
        </div>

        <div className='journal__entry-date-box'>
            <span>{noteDate.format('dddd')}</span>
            <h4>{noteDate.format('Do')}</h4>
        </div>
    </div>
  )
}
