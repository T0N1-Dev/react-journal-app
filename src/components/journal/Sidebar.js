import React from 'react'
import { JournalEntries } from './JournalEntries'
import { useDispatch, useSelector } from 'react-redux'
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';

export const Sidebar = () => {

    const dispatch = useDispatch();
    const { name } = useSelector(state => state.auth);

    const handleLogout = () => {
        dispatch( startLogout() );
    }

    const handleAddNew = () => {
        dispatch( startNewNote() );
    }

  return (
    <aside className='journal__sidebar'>
        <div className='journal__sidebar-navbar'>
            <h3 className='mt-5'>&nbsp;&nbsp;
                <i className='far fa-moon'></i>&nbsp;&nbsp;
                <span>{ name }</span>
            </h3>

            <button className='btn mt-5' onClick={ handleLogout }>
                Logout
            </button>
        </div>

        <div 
        className='journal__new-entry'
        onClick={ handleAddNew }
        
        >
            <i className='far fa-calendar-plus fa-5x mb-1'></i>
            <p>
                New entry
            </p>
        </div>

        <JournalEntries />
        
    </aside>
  )
}
