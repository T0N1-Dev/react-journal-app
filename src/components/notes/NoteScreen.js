import React, { useEffect, useRef } from 'react'
import { NotesAppBar } from './NotesAppBar'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../../hooks/useForm';
import { activeNote, startDeleting } from '../../actions/notes';

export const NoteScreen = () => {

  const dispatch = useDispatch();

  const { active: noteActive } = useSelector( state => state.notes );
  const [ formValues, handleInputChange, reset ] = useForm( noteActive );
  const { body, title } = formValues;

  const activeId = useRef( noteActive.id );

  useEffect(() => {
    if ( noteActive.id !== activeId.current ) {
      reset( noteActive );
      activeId.current = noteActive.id;
    }
  }, [noteActive, reset]);

  const handleDelete = () => {
    dispatch( startDeleting(activeId.current) );
  }

  useEffect(() => {
    dispatch( activeNote(formValues.id, {...formValues}));
    
  }, [formValues, dispatch]);
  

  return (
    <div 
      className='notes__main-content animate__animated animate__fadeIn animate__faster'
    >
        <NotesAppBar />

        <div className='notes__content'>

          <input 
            type='text'
            placeholder='Some awesome title'
            className='notes__title-input'
            name='title'
            value={ title }
            onChange={handleInputChange}
          />

          <textarea
            placeholder='What appened today'
            className='notes__textarea'
            name='body'
            value={ body }
            onChange={handleInputChange}
          ></textarea>

          {
            (noteActive.url)
            &&
            (<div className='notes__image'>
              <img src={ noteActive.url } alt={ noteActive.title } />
            </div>)
          }
        </div>

        <button
          className='btn btn-danger'
          onClick={ handleDelete }
        > 
        Delete
        </button>
    </div>
  )
}
