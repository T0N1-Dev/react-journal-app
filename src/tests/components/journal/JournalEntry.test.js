import { fireEvent, render } from "@testing-library/react";
import { JournalEntry } from "../../../components/journal/JournalEntry";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { activeNote } from "../../../actions/notes";

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()
}));

describe('Tests in <JournalEntry />', () => {

    let store;
    let testContainer;
    const note = {
        id: 10,
        date: 0,
        title: 'Hello',
        body: 'World',
        url: 'https://someurl.com'
    }

    beforeEach(() => {
        store = configureStore({
            reducer: {
            },
            preloadedState: {
            }
        })

        store.dispatch = jest.fn();

        const {container} = render(
            <Provider store={store}>
                <JournalEntry {...note} />
            </Provider>
        );

        testContainer = container;
    });

    test('should match snapshot', () => {
        expect(testContainer).toMatchSnapshot();
    })

    test('should call activeNote', () => {
        const noteActiveDiv = testContainer.querySelector('.journal__entry');
        fireEvent.click(noteActiveDiv);
        const { id, ...noteWithoutId } = note;

        expect(activeNote).toHaveBeenCalled();
        expect(activeNote).toHaveBeenCalledWith(id, noteWithoutId);
    })
});