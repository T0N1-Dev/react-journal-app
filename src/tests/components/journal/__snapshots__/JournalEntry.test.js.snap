// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <JournalEntry /> should match snapshot 1`] = `
<div>
  <div
    class="journal__entry pointer animate__animated animate__backInLeft animate__faster"
  >
    <div
      class="journal_entry-picture"
      style="background-size: cover; background-image: url(https://someurl.com);"
    />
    <div
      class="journal_entry-body"
    >
      <p
        class="journal__entry-title"
      >
        Hello
      </p>
      <p
        class="journal__entry-content"
      >
        World
      </p>
    </div>
    <div
      class="journal__entry-date-box"
    >
      <span>
        Wednesday
      </span>
      <h4>
        31st
      </h4>
    </div>
  </div>
</div>
`;
